pkg/importer/dataStream.go

```diff
@@ -172,97 +178,155 @@ func CopyImage(dest, endpoint, accessKey, secKey string) error {
+	glog.V(Vadmin).Infof("create the initial Reader based on the url's %q scheme", d.Url.Scheme)
+	r, err := d.dataStreamSelector()
 	if err != nil {
 		return errors.WithMessage(err, "could not get data reader")
 	}
-	d.Readers = append(d.Readers, rdr)
+	// append source reader to datastream reader stack
+	d.Readers = append(d.Readers, r)
 
-	// build slice of compression/archive extensions in right-to-left order
-	exts := []string{}
-	fn := d.Url.Path
-	for image.IsSupportedCompressArchiveType(fn) {
-		ext := strings.ToLower(filepath.Ext(fn))
-		exts = append(exts, ext)
-		fn = strings.TrimSuffix(fn, ext)
-	}
-	// create decompress/un-archive Readers
-	glog.V(Vdebug).Infof("constructReaders: checking compressed and/or archive for file %q\n", d.Url.Path)
-	for _, ext := range exts {
-		switch ext {
-		case image.ExtGz:
-			rdr, err = GzReader(rdr.Rdr)
-		case image.ExtTar:
-			rdr, err = TarReader(rdr.Rdr)
-		case image.ExtXz:
-			rdr, err = XzReader(rdr.Rdr)
+	// loop through all supported file formats until we do not find a header we recognize
+	knownHdrs := image.CopyKnownHdrs() // need local copy since keys are removed
+	glog.V(Vdebug).Infof("constructReaders: checking compression and archive formats: %s\n", d.Url.Path)
+	for {
+		hdr, err := d.matchHeader(&knownHdrs)
+		if err != nil {
+			return errors.WithMessage(err, "could not process image header")
+		}
+		// append multi-reader so that the header data is re-read by subsequent readers
+		d.Readers = append(d.Readers, Reader{RdrType: RdrMulti, Rdr: ioutil.NopCloser(io.MultiReader(bytes.NewReader(d.buf), d.topReader()))})
+		if hdr == nil {
+			break // done processing headers, we have the orig source file
 		}
+		glog.V(Vadmin).Infof("found header of type %q\n", hdr.Format)
+
+		// create format-specific reader and append it to dataStream Readers slice
+		err = d.fileFormatSelector(hdr)
 		if err != nil {
-			return errors.WithMessage(err, "could not get compression reader")
+			return errors.WithMessage(err, "could not create compression/unarchive reader")
 		}
-		d.Readers = append(d.Readers, rdr)
 	}
+	if len(d.Readers) <= 2 { // 1st rdr is source, 2nd rdr is multi-rdr, >2 means we have additional formats
+		glog.V(Vdebug).Infof("constructReaders: no headers found for file %q\n", d.Url.Path)
+	}
+	glog.V(Vadmin).Infof("done processing %q headers\n", d.Url.Path)
+	return nil
+}
+
+// Return the top-level io.ReadCloser from the receiver Reader "stack".
+func (d dataStream) topReader() io.ReadCloser {
+	return d.Readers[len(d.Readers)-1].Rdr
+}
+
+// Based on the passed in header, return the format-specific reader.
+func (d *dataStream) fileFormatSelector(hdr *image.Header) error {
+	switch hdr.Format {
+	case "gz":
+		return d.gzReader()
+	case "qcow2":
+		return d.qcow2NopReader(hdr)
+	case "tar":
+		return d.tarReader()
+	case "xz":
+		return d.xzReader()
+	default:
+		return errors.Errorf("mismatch between supported file formats and this header type: %q", hdr.Format)
+	}
+}
```

Comment:
creating a new datastream Reader type with the format reader and returning that?  Then appending that at the caller's scope.